File: 2358-maximum-number-of-groups-entering-a-competition/2358-maximum-number-of-groups-entering-a-competition.java
class Solution {
    public int maximumGroups(int[] A) {
        int left = 0, right = 1000, n = A.length;
        while (left < right) {
            int k = (left + right + 1) / 2;
            if (k * (k + 1) / 2 > n) {
                right = k - 1;
            } else {
                left = k;
            }
        }
        return left;
    }
}

File: 2358-maximum-number-of-groups-entering-a-competition/2358-maximum-number-of-groups-entering-a-competition.cpp
class Solution
{
    public:
        int maximumGroups(vector<int> &grades)
        {
            long long n = grades.size();
            long long low = 1, high = n, maxGrps = n + 1;
            while (low <= high)
            {
                long long mid = (high + low) / 2;
                long long sum = mid *(mid + 1) / 2;
                if (sum <= n)
                {
                    maxGrps = mid;
                    low = mid + 1;
                }
                else
                {
                    high = mid - 1;
                }
            }
            return maxGrps;
        }
};